File: code/python/index_ranges2.py
#!/usr/bin/env python

import json
import time
import datetime
import sys
import xapian
from parsecsv import parse_csv_file

def numbers_from_string(s):
    """Find all numbers in a string."""
    numbers = []
    in_number = False
    while len(s) > 0:
        next_in_number = (s[0].isdigit() or s[0]=='.')
        if next_in_number != in_number:
            in_number = not in_number
            if in_number:
                numbers.append(s[0])
        else:
            if in_number:
                numbers[-1] = numbers[-1] + s[0]
        s = s[1:]
    # fix up leading or trailing '.'
    for index, number in enumerate(numbers):
        if number[0]=='.':
            number = '0.' + number[1:]
        if number[-1]=='.':
            number += '0'
        numbers[index] = float(number)
    return numbers


def index(datapath, dbpath):
    # Create or open the database we're going to be writing to.
    db = xapian.WritableDatabase(dbpath, xapian.DB_CREATE_OR_OPEN)

    # Set up a TermGenerator that we'll use in indexing.
    termgenerator = xapian.TermGenerator()
    termgenerator.set_stemmer(xapian.Stem("en"))

    for fields in parse_csv_file(datapath, 'utf-8'):
        # 'fields' is a dictionary mapping from field name to value.
        admitted = fields.get('admitted', None)
        if admitted is None:
            print "Couldn't process", fields
            continue
        else:
            # Date (order) -- we use the order as our identifier
            pieces = admitted.split('(', 1)
            admitted = pieces[0]
            admitted.strip()
            try:
                admitted = time.strptime(
                    admitted,
                    "%B %d, %Y ",
                    )
                # now a struct_time, convert to YYYYMMDD
                admitted = "%4.4i%2.2i%2.2i" % (
                    admitted[0],
                    admitted[1],
                    admitted[2],
                    )
            except ValueError:
                print "couldn't parse admitted '%s'" % admitted
                admitted = None

            order = pieces[1]
            if order[-1] == ')':
                order = order[:-1]
            if order[-2:] in ('st', 'nd', 'rd', 'th'):
                order = order[:-2]
            order.strip()
        population = fields.get('population', None)
        if population is not None:
            # Population-comma-formatted (comment) extra
            pieces = population.split('(', 1)
            population = pieces[0].replace(',', '')
            population.strip()
            try:
                population = int(population)
            except ValueError:
                population = None
        
        # Note that the following code isn't used in explaining
        # range queries at all. I thought I was going to use it,
        # but ended up not bothering. It may prove useful in
        # explaining geosearch at some point in the future.
        def middle_coord(text):
            """
            In the form <start> to <end> with both extents being in
            degrees and minutes as N/S/W/E. S and W thus need to be
            negated as we only care about N/E.
            """
            if text is None:
                return None
            pieces = text.split(' to ', 1)
            #print "coords", pieces
            start, end = map(numbers_from_string, pieces)
            def tuple_to_float(numbers):
                divisor = 1
                result = 0
                while len(numbers) > 0:
                    result += float(numbers[0]) / divisor
                    numbers = numbers[1:]
                    divisor = divisor * 60
                return result
            start = tuple_to_float(start)
            end = tuple_to_float(end)
            if pieces[0][-1] in ('S', 'W'):
                start = -start
            if pieces[1][-1] in ('S', 'W'):
                end = -end
            return (start + end) / 2
        midlat = middle_coord(fields.get('latitude', None))
        midlon = middle_coord(fields.get('longitude', None))

        # We make a document and tell the term generator to use this.
        doc = xapian.Document()
        termgenerator.set_document(doc)

        # Index each field with a suitable prefix.
        name = fields.get('name', u'')
        description = fields.get('description', u'')
        motto = fields.get('motto', u'')
        # index each field with a suitable prefix
        termgenerator.index_text(name, 1, 'S')
        termgenerator.index_text(description, 1, 'XD')
        termgenerator.index_text(motto, 1, 'XM')

        # Index fields without prefixes for general search.
        termgenerator.index_text(name)
        termgenerator.increase_termpos()
        termgenerator.index_text(description)
        termgenerator.increase_termpos()
        termgenerator.index_text(motto)

        # Store all the fields for display purposes.
        doc.set_data(json.dumps(fields))

        # add document values
        if admitted is not None:
            doc.add_value(1, xapian.sortable_serialise(int(admitted[:4])))
            doc.add_value(2, admitted) # YYYYMMDD
        if population is not None:
            doc.add_value(3, xapian.sortable_serialise(population))

        # we use the identifier to ensure each object ends up
        # in the database only once no matter how many times
        # we run the indexer
        idterm = u"Q" + order
        doc.add_boolean_term(idterm)
        db.replace_document(idterm, doc)

index(datapath = sys.argv[1], dbpath = sys.argv[2])

File: code/c++/index_ranges2.cc
#include <xapian.h>

#include <cstdlib>
#include <fstream>
#include <iostream>
#include <string>
#include <vector>

#include "support.h"

using namespace std;

void index(const string & datapath, const string & dbpath)
{
    // Hardcode field offsets for simplicity.
    const size_t FIELD_NAME = 0;
    const size_t FIELD_ADMITTED = 2;
    const size_t FIELD_ORDER = 3;
    const size_t FIELD_POPULATION = 4;
    const size_t FIELD_MOTTO = 7;
    const size_t FIELD_DESCRIPTION = 8;

    // Create or open the database we're going to be writing to.
    Xapian::WritableDatabase db(dbpath, Xapian::DB_CREATE_OR_OPEN);

    // Set up a TermGenerator that we'll use in indexing.
    Xapian::TermGenerator termgenerator;
    termgenerator.set_stemmer(Xapian::Stem("en"));

    ifstream csv(datapath.c_str());
    vector<string> fields;
    csv_parse_line(csv, fields);

    // Check the CSV header line matches our hard-code offsets.
    if (fields.at(FIELD_NAME) != "name" ||
	fields.at(FIELD_ADMITTED) != "admitted" ||
	fields.at(FIELD_ORDER) != "order" ||
	fields.at(FIELD_POPULATION) != "population" ||
	fields.at(FIELD_MOTTO) != "motto" ||
	fields.at(FIELD_DESCRIPTION) != "description") {
	// The CSV format doesn't match what we expect.
	cerr << "CSV format has changed!" << endl;
	exit(1);
    }

    while (csv_parse_line(csv, fields)) {
	// 'fields' is a vector mapping from field number to value.
	// We look up fields with the 'at' method so we get an exception
	// if that field isn't set.
	const string & name = fields.at(FIELD_NAME);
	const string & description = fields.at(FIELD_DESCRIPTION);
	const string & motto = fields.at(FIELD_MOTTO);
	const string & admitted = fields.at(FIELD_ADMITTED);
	const string & population = fields.at(FIELD_POPULATION);
	const string & order = fields.at(FIELD_ORDER);

	// We make a document and tell the term generator to use this.
	Xapian::Document doc;
	termgenerator.set_document(doc);

// Start of example code.
	// Index each field with a suitable prefix.
	termgenerator.index_text(name, 1, "S");
	termgenerator.index_text(description, 1, "XD");
	termgenerator.index_text(motto, 1, "XM");

	// Index fields without prefixes for general search.
	termgenerator.index_text(name);
	termgenerator.increase_termpos();
	termgenerator.index_text(description);
	termgenerator.increase_termpos();
	termgenerator.index_text(motto);

	// Add document values.
	if (!admitted.empty()) {
	    doc.add_value(1, Xapian::sortable_serialise(atoi(admitted.substr(0, 4).c_str())));
	    doc.add_value(2, admitted); // YYYYMMDD
	}
	if (!population.empty()) {
	    doc.add_value(3, Xapian::sortable_serialise(atoi(population.c_str())));
	}
// End of example code.

	// Store all the fields for display purposes.
	doc.set_data(name + "\n" + description + "\n" + motto + "\n" +
		     admitted + "\n" + population + "\n" + order);

	// We use the order to ensure each object ends up in the
	// database only once no matter how many times we run the
	// indexer.
	string idterm = "Q" + order;
	doc.add_boolean_term(idterm);
	db.replace_document(idterm, doc);
    }
}

int main(int argc, char** argv) {
    if (argc != 3) {
	cerr << "Usage: " << argv[0] << " DATAPATH DBPATH" << endl;
	return 1;
    }
    index(argv[1], argv[2]);
}
